app/src/main/java/com/willowtree/vocable/presets/AddToCategoryPickerFragment.kt

```diff
@@ -0,0 +1,162 @@
+                0 -> {
+                    binding.categoryHolder.setCurrentItem(
+                        categoriesAdapter.itemCount - 1,
+                        true
+                    )
+                }
+                else -> {
+                    binding.categoryHolder.setCurrentItem(currentPosition - 1, true)
+                }
+            }
+        }
+
+        binding.categoryHolder.registerOnPageChangeCallback(object :
+            ViewPager2.OnPageChangeCallback() {
+            override fun onPageSelected(position: Int) {
+                val pageNum = position % categoriesAdapter.numPages + 1
+                binding.categoryPageNumber.text = getString(R.string.phrases_page_number, pageNum, categoriesAdapter.numPages)
+
+                val activity = requireActivity()
+                if (activity is MainActivity) {
+                    activity.resetAllViews()
+                }
+            }
+        })
+
+        addToCategoryPickerViewModel = ViewModelProviders.of(
+            requireActivity(),
+            BaseViewModelFactory()
+        ).get(AddToCategoryPickerViewModel::class.java)
+
+        subscribeToViewModel()
+
+        with(addToCategoryPickerViewModel) {
+            buildCategoryList(phraseText)
+        }
+    }
+
+    private fun subscribeToViewModel() {
+        addToCategoryPickerViewModel.categoryMap.observe(viewLifecycleOwner, Observer {
+            it?.let { map ->
+                with(ArrayList(map.keys)) {
+                    val categoriesExist = this.size != 0
+
+                    binding.emptyStateTitle.isVisible = !categoriesExist
+                    binding.emptyStateText.isVisible = !categoriesExist
+                    binding.categoryPagerBackButton.isEnabled = categoriesExist
+                    binding.categoryPagerForwardButton.isEnabled = categoriesExist
+
+                    // empty state
+                    if (this.size == 0) {
+                        binding.categoryPageNumber.text = getString(R.string.phrases_page_number, 1, 1)
+                    } else {
+                        binding.categoryHolder.apply {
+                            isSaveEnabled = false
+                            adapter = categoriesAdapter
+                            categoriesAdapter.setItems(this@with)
+
+                            // Move adapter to middle so user can scroll both directions
+                            val middle = categoriesAdapter.itemCount / 2
+                            if (middle % categoriesAdapter.numPages == 0) {
+                                setCurrentItem(middle, false)
+                            } else {
+                                val mod = middle % categoriesAdapter.numPages
+                                setCurrentItem(
+                                    middle + (categoriesAdapter.numPages - mod),
+                                    false
+                                )
+                            }
+                        }
```

Comment:
So this is probably another example of where extracting all of this logic out into its own benefits the code base bc it makes it more readable which also lessens the chances for mistakes. It'd also be easier to find em too. There are a couple of KTX extension functions that help out with this too.

So usual on my projects we'd have a method just like your `subscribeToViewModel()` except ours is a lil smaller and intentional. Each LiveData is being observed and firing off their own methods to handle.
```
private fun observeViewModel() = viewModel.apply {
        recommendation.observe(viewLifecycleOwner, ::bindRecommendation)
        feedback.observe(viewLifecycleOwner, ::bindFeedback)
        error.observe(viewLifecycleOwner, ::handleError)
    }

private fun bindRecommendation(recommendation: LolaRecommendation) {
    // Do thangs
}
```